### PR TITLE
HDDS-7760. Let ContainerDataConstructor extend SafeConstructor

### PR DESCRIPTION
**Pull request by rohit-kb**
## What changes were proposed in this pull request?

Finding and replacing Constructor() class of snakeyaml with SafeConstructor() due to CVE-2022-1471

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-7760

### CODE REVIEW

**Review by adoroszlai — COMMENTED** [review]
Thanks @rohit-kb for working on this.

I think we also need to update `new Yaml` calls in a few places, passing `SafeConstructor` instance:

https://github.com/apache/ozone/blob/a80c6b1c73b01660e067ba1c30e055522fd1fdfb/hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/scm/net/NodeSchemaLoader.java#L230

https://github.com/apache/ozone/blob/a80c6b1c73b01660e067ba1c30e055522fd1fdfb/hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/helpers/DatanodeIdYaml.java#L59

https://github.com/apache/ozone/blob/a80c6b1c73b01660e067ba1c30e055522fd1fdfb/hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/helpers/DatanodeIdYaml.java#L74

https://github.com/apache/ozone/blob/a80c6b1c73b01660e067ba1c30e055522fd1fdfb/hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/TestMiniOzoneCluster.java#L191

**Comment by rohit-kb** [comment]
thanks @adoroszlai, just want to a little about the role of SafeConstructor instance in Yaml object creations.

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/helpers/DatanodeIdYaml.java [hunk 1]
```diff
@@ -56,7 +59,7 @@ public static void createDatanodeIdFile(DatanodeDetails datanodeDetails,
     DumperOptions options = new DumperOptions();
     options.setPrettyFlow(true);
     options.setDefaultFlowStyle(DumperOptions.FlowStyle.FLOW);
-    Yaml yaml = new Yaml(options);
+    Yaml yaml = new Yaml(new SafeConstructor(new LoaderOptions()), new Representer(options), options);
```
Comment (adoroszlai) [inline hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/helpers/DatanodeIdYaml.java:62]:
```
hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/helpers/DatanodeIdYaml.java
 62: Line is longer than 80 characters (found 102).
```

https://github.com/rohit-kb/ozone/actions/runs/3882072979/jobs/6621789972#step:6:427

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/scm/net/NodeSchemaLoader.java [hunk 2]
```diff
@@ -227,7 +229,7 @@ private NodeSchemaLoadResult loadSchemaFromYaml(InputStream schemaFile) {
     NodeSchemaLoadResult finalSchema;
 
     try {
-      Yaml yaml = new Yaml();
+      Yaml yaml = new Yaml(new SafeConstructor(new LoaderOptions()));
```
Comment (adoroszlai) [inline hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/scm/net/NodeSchemaLoader.java:232]: Looks like this causes test failures:

```
 Tests run: 5, Failures: 4, Errors: 0, Skipped: 0, Time elapsed: 0.268 s <<< FAILURE! - in org.apache.hadoop.hdds.scm.net.TestYamlSchemaLoader
```

https://github.com/rohit-kb/ozone/actions/runs/3882072979/jobs/6621790406#step:6:584

Can you please check?

**Review by adoroszlai — COMMENTED** [review]
Thanks @rohit-kb for updating the patch.

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/scm/net/NodeSchemaLoader.java (hunk 2 above)
Comment (rohit-kb) [reply hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/scm/net/NodeSchemaLoader.java:232]: thanks for the review, looking into it

**Review by adoroszlai — APPROVED** [review]

**Pull request merged**